extensions/sql-database-projects/src/controllers/databaseProjectTreeViewProvider.ts

```diff
@@ -0,0 +1,132 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import * as nls from 'vscode-nls';
+
+import { SqlDatabaseProjectItem } from './databaseProjectTreeItem';
+import { Deferred } from '../common/promise';
+
+
+const localize = nls.loadMessageBundle();
+
+export class SqlDatabaseProjectTreeViewProvider implements vscode.TreeDataProvider<SqlDatabaseProjectItem> {
+	private _onDidChangeTreeData: vscode.EventEmitter<SqlDatabaseProjectItem | undefined> = new vscode.EventEmitter<SqlDatabaseProjectItem | undefined>();
+	readonly onDidChangeTreeData: vscode.Event<SqlDatabaseProjectItem | undefined> = this._onDidChangeTreeData.event;
+
+	private _initializeDeferred: Deferred<void> = new Deferred<void>();
+
+	private roots: SqlDatabaseProjectItem[] = [];
+
+	constructor() {
+		this._initializeDeferred = new Deferred<void>();
+		this.initialize();
+	}
+
+	async initialize() {
```

Comment:
if there a reason you need to do this? in general deferred promises are a anti-pattern (why its not built into javascript). So if you don't need to do it, you should avoid it.

Reply:
Yeah, I'd suggest removing this until you actually need to do async operations (and even then try to avoid this being necessary)

---
In reply to: [356820406](https://github.com/microsoft/azuredatastudio/pull/8639#discussion_r356820406) [](ancestors = 356820406)

Reply:
Nope; just copied the shape of a new extension.  Removed.